app/lib/frontend/backend.dart

```diff
@@ -159,6 +162,128 @@ class Backend {
+      if (invite != null && !invite.isExpired()) {
+        invite.lastNotified = now;
+        invite.notificationCount++;
+        tx.queueMutations(inserts: [invite]);
+        await tx.commit();
+        return new InviteStatus(urlNonce: invite.urlNonce);
+      }
+
+      // Reset old or create new invite.
+      invite ??= models.PackageInvite()
+        ..parentKey = pkgKey
+        ..id = inviteId
+        ..type = type
+        ..recipientEmail = recipientEmail;
+      final urlNonce =
+          new List.generate(25, (i) => _random.nextInt(36).toRadixString(36))
+              .join();
+      invite
+        ..urlNonce = urlNonce
+        ..fromEmail = fromEmail
+        ..created = now
+        ..expires = now.add(Duration(days: 1))
+        ..notificationCount = 1
+        ..lastNotified = now;
+      tx.queueMutations(inserts: [invite]);
+      await tx.commit();
+      return new InviteStatus(urlNonce: invite.urlNonce);
+    }) as InviteStatus;
+  }
+
+  /// Confirm the invite and return the Datastore entry object if successful,
+  /// otherwise returns null.
+  Future<models.PackageInvite> confirmPackageInvite({
+    @required String packageName,
+    @required String type,
+    @required String recipientEmail,
+    @required String urlNonce,
+  }) async {
+    final inviteId = models.PackageInvite.createId(type, recipientEmail);
+    final pkgKey = db.emptyKey.append(models.Package, id: packageName);
+    final inviteKey = pkgKey.append(models.PackageInvite, id: inviteId);
+    return await db.withTransaction((tx) async {
+      final list = await tx.lookup([inviteKey]);
+      models.PackageInvite invite = list.single;
+
+      // Invite entry does not exists.
+      if (invite == null) {
+        await tx.rollback();
+        return null;
+      }
+
+      // Invite entry has expired.
+      if (invite.isExpired()) {
+        tx.queueMutations(deletes: [inviteKey]);
+        await tx.commit();
+        return null;
+      }
+
+      // urlNonce check: whether the invite has been updated.
+      if (invite.urlNonce != urlNonce) {
+        await tx.rollback();
+        return null;
+      }
+
+      // Invite already confirmed.
+      if (invite.confirmed != null) {
+        await tx.rollback();
+        return invite;
+      }
+
+      // Confirming invite.
+      invite.confirmed = new DateTime.now().toUtc();
+      tx.queueMutations(inserts: [invite]);
+      await tx.commit();
+      return invite;
+    }) as models.PackageInvite;
+  }
+
+  /// Removes obsolete (== expired more than a day ago) invites from Datastore.
+  Future deleteObsoleteInvites() async {
+    final query = db.query<models.PackageInvite>()
+      ..filter('expires <', DateTime.now().subtract(Duration(days: 1)));
+    await for (var invite in query.run()) {
+      db.commit(deletes: [invite.key]);
+    }
+  }
+}
+
+/// The status of an invite after being created or updated.
+class InviteStatus {
+  final String urlNonce;
+  final Duration delayDuration;
```

Comment:
Should this be a timestamp of the actual availability (creationTime + delayDuration)?

Reply:
I was thinking that the message to the client would be "You need to wait x more minutes before re-sending the invite.", but as they are almost equivalent, we can use timestamp too. Which do you prefer?

Reply:
Ah - I read the code wrong - it is already calculated relative to now.
Then go with whatever suits you better.